core/src/test/java/com/datastax/dse/driver/internal/core/cql/reactive/ReactiveResultSetSubscriptionTest.java

```diff
@@ -145,4 +145,31 @@ public void should_report_error_on_intermediary_page() {
     assertThat(wasAppliedSubscriber.getElements()).hasSize(1).containsExactly(true);
     assertThat(wasAppliedSubscriber.getError()).isNull();
   }
+
+  @Test
+  public void should_handle_empty_non_final_pages() {
+    CompletableFuture<AsyncResultSet> future1 = new CompletableFuture<>();
+    CompletableFuture<AsyncResultSet> future2 = new CompletableFuture<>();
+    CompletableFuture<AsyncResultSet> future3 = new CompletableFuture<>();
+    MockAsyncResultSet page1 = new MockAsyncResultSet(10, future2);
+    MockAsyncResultSet page2 = new MockAsyncResultSet(0, future3);
+    MockAsyncResultSet page3 = new MockAsyncResultSet(0, null);
+    TestSubscriber<ReactiveRow> mainSubscriber = new TestSubscriber<>(1);
+    TestSubscriber<ColumnDefinitions> colDefsSubscriber = new TestSubscriber<>();
+    TestSubscriber<ExecutionInfo> execInfosSubscriber = new TestSubscriber<>();
+    TestSubscriber<Boolean> wasAppliedSubscriber = new TestSubscriber<>();
+    ReactiveResultSetSubscription<AsyncResultSet> subscription =
+        new ReactiveResultSetSubscription<>(
+            mainSubscriber, colDefsSubscriber, execInfosSubscriber, wasAppliedSubscriber);
+    mainSubscriber.onSubscribe(subscription);
+    subscription.start(() -> future1);
+    future1.complete(page1);
+    future2.complete(page2);
+    // emulate backpressure
+    subscription.request(1);
+    future3.complete(page3);
+    subscription.request(Long.MAX_VALUE);
+    mainSubscriber.awaitTermination();
+    assertThat(mainSubscriber.isCompleted()).isTrue();
+  }
```

Comment:
It's a minor thing but I prefer to not change `TestSubscriber` and rewrite the test as suggested.

```suggestion
  public void should_handle_empty_non_final_pages() {
    CompletableFuture<AsyncResultSet> future1 = new CompletableFuture<>();
    CompletableFuture<AsyncResultSet> future2 = new CompletableFuture<>();
    CompletableFuture<AsyncResultSet> future3 = new CompletableFuture<>();
    MockAsyncResultSet page1 = new MockAsyncResultSet(1, future2);
    MockAsyncResultSet page2 = new MockAsyncResultSet(0, future3); // empty
    MockAsyncResultSet page3 = new MockAsyncResultSet(1, null);
    TestSubscriber<ReactiveRow> mainSubscriber = new TestSubscriber<>();
    TestSubscriber<ColumnDefinitions> colDefsSubscriber = new TestSubscriber<>();
    TestSubscriber<ExecutionInfo> execInfosSubscriber = new TestSubscriber<>();
    TestSubscriber<Boolean> wasAppliedSubscriber = new TestSubscriber<>();
    ReactiveResultSetSubscription<AsyncResultSet> subscription =
        new ReactiveResultSetSubscription<>(
            mainSubscriber, colDefsSubscriber, execInfosSubscriber, wasAppliedSubscriber);
    mainSubscriber.onSubscribe(subscription);
    subscription.start(() -> future1);
    future1.complete(page1);
    future2.complete(page2);
    // row 1 should be served from page1
    assertThat(mainSubscriber.getElements()).hasSize(1);
    // page1 and page2 should have been discarded, queue should be empty
    assertThat(subscription)
        .extracting("pages.elements")
        .satisfies(queue -> assertThat(((Queue<?>) queue)).isEmpty());
    // row 2 should be served from page3
    future3.complete(page3);
    mainSubscriber.awaitTermination();
    // should have completed normally
    assertThat(mainSubscriber.getError()).isNull();
    // should have received 2 rows
    List<Row> expected = new ArrayList<>(page1.currentPage());
    expected.addAll(page3.currentPage());
    assertThat(mainSubscriber.getElements()).hasSize(2).extracting("row").isEqualTo(expected);
  }
```

Reply:
If I rewrite test how you suggested it will pass with previous code. Because `future3.complete(page3);` will call `drain` -> `tryNext()` and code successfully complete.

I tried to rewrite this test without changing `TestSubscriber`, but didn't come up with anything

Reply:
I tried my changes before suggesting them. They work as expected, feel free to cherry-pick: 

https://github.com/adutra/java-driver/commit/466e5ddc276576495198be0d5a89a0e2d538c482

Reply:
@adutra Yes, it works. But it works for case with:
```
current = pages.peek();
        // if the next page is readily available,
        // serve its first row now, no need to wait
        // for the next drain.
        if (current != null && current.hasMoreRows()) {
          return current.nextRow();
        }
```
as well

Reply:
I meant if you change branch to `4.x` branch and try to use your test. It will work, but it should fail

Reply:
Ah right, now I understand what you had in mind. Please disregard my previous comments :-)